File: src/AetherNet.Space/SpaceBreadcrumbService.cs
// SPDX-License-Identifier: MIT

using System.Text.Json;
using System.Text.Json.Serialization;
using AetherNet.Constants;
using AetherNet.Protocol;
using AetherNet.Routing;
using AetherNet.Space.Models;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace AetherNet.Space;

/// <summary>
/// Wire payload for <see cref="PacketType.SpaceBreadcrumb"/> (40). Projects the
/// <see cref="SpaceBreadcrumb"/> model onto a byte-identical JSON shape: snake_case keys pinned by
/// <see cref="JsonPropertyNameAttribute"/>, the UTC creation time as a Unix-ms integer (not ISO-8601),
/// the category enum as a bare integer, and the Ed25519 signature as STANDARD base64. Field order:
/// content_hash, geo_hash, anchor_uhid, created_at_ms, ttl_hours, type, signature.
/// Byte-identity gate: fixtures/space/vectors.json.
/// </summary>
public sealed class SpaceBreadcrumbPayload
{
    [JsonPropertyName("content_hash")] public string ContentHash { get; set; } = string.Empty;
    [JsonPropertyName("geo_hash")] public string GeoHash { get; set; } = string.Empty;
    [JsonPropertyName("anchor_uhid")] public string AnchorUhid { get; set; } = string.Empty;
    [JsonPropertyName("created_at_ms")] public long CreatedAtMs { get; set; }
    [JsonPropertyName("ttl_hours")] public int TtlHours { get; set; }
    [JsonPropertyName("type")] public int Type { get; set; }
    [JsonPropertyName("signature")] public byte[] Signature { get; set; } = Array.Empty<byte>();

    public static SpaceBreadcrumbPayload FromBreadcrumb(SpaceBreadcrumb b) => new()
    {
        ContentHash = b.ContentHash,
        GeoHash = b.GeoHash,
        AnchorUhid = b.AnchorUhid,
        CreatedAtMs = new DateTimeOffset(DateTime.SpecifyKind(b.CreatedAtUtc, DateTimeKind.Utc)).ToUnixTimeMilliseconds(),
        TtlHours = b.TtlHours,
        Type = (int)b.Type,
        Signature = b.Signature,
    };

    public SpaceBreadcrumb ToBreadcrumb() => new()
    {
        ContentHash = ContentHash,
        GeoHash = GeoHash,
        AnchorUhid = AnchorUhid,
        CreatedAtUtc = DateTimeOffset.FromUnixTimeMilliseconds(CreatedAtMs).UtcDateTime,
        TtlHours = TtlHours,
        Type = (BreadcrumbType)(byte)Type,
        Signature = Signature,
    };
}

/// <summary>
/// Binds <see cref="PacketType.SpaceBreadcrumb"/> (40) to the mesh: broadcast a locally-dropped
/// breadcrumb, and surface inbound breadcrumbs via <see cref="BreadcrumbReceived"/> (the host pins them
/// into <c>ISpaceService</c>). Transport for the aether-space geo-pinned-notice extension.
/// </summary>
public interface ISpaceBreadcrumbService
{
    /// <summary>Raised when a breadcrumb arrives from a peer.</summary>
    event EventHandler<SpaceBreadcrumb>? BreadcrumbReceived;

    /// <summary>Flood a breadcrumb to mesh peers. Returns the number of peers it was delivered to.</summary>
    Task<int> BroadcastAsync(SpaceBreadcrumb breadcrumb, CancellationToken cancellationToken = default);

    /// <summary>Process an inbound <see cref="PacketType.SpaceBreadcrumb"/>. Returns false on wrong type or malformed payload.</summary>
    Task<bool> HandleAsync(MeshPacket packet, CancellationToken cancellationToken = default);
}

/// <inheritdoc />
public sealed class SpaceBreadcrumbService : ISpaceBreadcrumbService
{
    private static readonly JsonSerializerOptions JsonOptions = new();

    private readonly IMeshSender _sender;
    private readonly ILogger<SpaceBreadcrumbService> _logger;

    public event EventHandler<SpaceBreadcrumb>? BreadcrumbReceived;

    public SpaceBreadcrumbService(IMeshSender sender, ILogger<SpaceBreadcrumbService>? logger = null)
    {
        _sender = sender ?? throw new ArgumentNullException(nameof(sender));
        _logger = logger ?? NullLogger<SpaceBreadcrumbService>.Instance;
    }

    /// <inheritdoc />
    public async Task<int> BroadcastAsync(SpaceBreadcrumb breadcrumb, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(breadcrumb);
        var payload = SpaceBreadcrumbPayload.FromBreadcrumb(breadcrumb);
        var packet = new MeshPacket
        {
            Type = PacketType.SpaceBreadcrumb,
            SourceUhid = _sender.LocalUhid,
            DestinationUhid = "*",
            Ttl = ProtocolConstants.DefaultTtl,
            Payload = JsonSerializer.SerializeToUtf8Bytes(payload, JsonOptions),
        };
        var delivered = await _sender.BroadcastAsync(packet, cancellationToken).ConfigureAwait(false);
        _logger.LogDebug("SpaceBreadcrumb {Hash}@{Geo} broadcast to {N} peers", breadcrumb.ContentHash, breadcrumb.GeoHash, delivered);
        return delivered;
    }

    /// <inheritdoc />
    public Task<bool> HandleAsync(MeshPacket packet, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(packet);
        if (packet.Type != PacketType.SpaceBreadcrumb)
            return Task.FromResult(false);

        SpaceBreadcrumbPayload? body;
        try
        {
            body = JsonSerializer.Deserialize<SpaceBreadcrumbPayload>(packet.Payload, JsonOptions);
        }
        catch (JsonException ex)
        {
            _logger.LogDebug(ex, "SpaceBreadcrumb from {Source}: malformed payload — dropped", packet.SourceUhid);
            return Task.FromResult(false);
        }
        if (body is null || string.IsNullOrEmpty(body.ContentHash))
            return Task.FromResult(false);

        BreadcrumbReceived?.Invoke(this, body.ToBreadcrumb());
        return Task.FromResult(true);
    }
}

File: swift/Sources/AetherNetProtocol/Space/SpaceBreadcrumbService.swift
// SPDX-License-Identifier: MIT

// Wire binding for PacketType.spaceBreadcrumb (40) — the transport half of the
// aether-space geo-pinned-noticeboard extension. A thin actor that BROADCASTS a
// locally-dropped breadcrumb and surfaces inbound breadcrumbs via a callback (the
// host pins them into its SpaceServiceProtocol). Port of the C# reference
// (AetherNet.Space.SpaceBreadcrumbService).
//
// Wire payload: JSON, snake_case keys, field order
//   content_hash, geo_hash, anchor_uhid, created_at_ms, ttl_hours, type, signature
// created_at_ms is the UTC creation time as a bare Unix-ms integer (Int64), ttl_hours
// and type are bare integers (type = BreadcrumbType raw value: Notice=0, Emergency=1,
// Event=3), signature is STANDARD base64 (empty string when unsigned).
// Byte-identity gate: fixtures/space/vectors.json.

import Foundation

/// Binds ``PacketType/spaceBreadcrumb`` (40) to the mesh: broadcast a locally-dropped
/// breadcrumb, and surface inbound breadcrumbs via ``onBreadcrumbReceived`` (the host
/// pins them into its ``SpaceServiceProtocol``). Transport for the aether-space
/// geo-pinned-notice extension. Mirrors C# `SpaceBreadcrumbService`.
public actor SpaceBreadcrumbService {
    private let sender: any MeshSender

    /// Raised when a breadcrumb arrives from a peer.
    public var onBreadcrumbReceived: (@Sendable (SpaceBreadcrumb) -> Void)?

    public init(sender: any MeshSender) {
        self.sender = sender
    }

    public func setOnBreadcrumbReceived(_ callback: (@Sendable (SpaceBreadcrumb) -> Void)?) {
        onBreadcrumbReceived = callback
    }

    /// Flood a breadcrumb to mesh peers. Returns the number of peers it was delivered to.
    @discardableResult
    public func broadcast(_ breadcrumb: SpaceBreadcrumb) async -> Int {
        let packet = MeshPacket(
            type: .spaceBreadcrumb,
            sourceUhid: sender.localUhid,
            destinationUhid: "*",
            ttl: ProtocolConstants.defaultTtl,
            payload: encodeSpaceBreadcrumbWire(breadcrumb)
        )
        return await sender.broadcast(packet)
    }

    /// Process an inbound ``PacketType/spaceBreadcrumb``. Fires ``onBreadcrumbReceived``.
    /// Returns false on the wrong packet type or a malformed / empty-content-hash payload.
    @discardableResult
    public func handle(_ packet: MeshPacket) async -> Bool {
        guard packet.type == .spaceBreadcrumb else { return false }
        guard let crumb = parseSpaceBreadcrumbWire(packet.payload), !crumb.contentHash.isEmpty else {
            return false
        }
        onBreadcrumbReceived?(crumb)
        return true
    }
}

// ─── SpaceBreadcrumb wire (PacketType 40) ───
//
// Foundation's JSONEncoder does NOT emit keys in a deterministic declaration order — with
// 3+ fields it hash-reorders them, breaking cross-language byte-identity. So the wire JSON
// is built BY HAND in the exact field order (content_hash, geo_hash, anchor_uhid,
// created_at_ms, ttl_hours, type, signature), mirroring the other language ports. Decode
// uses JSONDecoder (order-independent).

/// Decode-only mirror of the wire shape. Only used to parse inbound payloads.
private struct SpaceBreadcrumbWire: Codable {
    let content_hash: String
    let geo_hash: String
    let anchor_uhid: String
    let created_at_ms: Int64
    let ttl_hours: Int
    let type: Int
    let signature: String
    private enum CodingKeys: String, CodingKey {
        case content_hash, geo_hash, anchor_uhid, created_at_ms, ttl_hours, type, signature
    }
}

private func jsonEscaped(_ s: String) -> String {
    var out = "\""
    for scalar in s.unicodeScalars {
        switch scalar {
        case "\"": out += "\\\""
        case "\\": out += "\\\\"
        case "\n": out += "\\n"
        case "\r": out += "\\r"
        case "\t": out += "\\t"
        default:
            if scalar.value < 0x20 { out += String(format: "\\u%04x", scalar.value) }
            else { out.unicodeScalars.append(scalar) }
        }
    }
    out += "\""
    return out
}

/// Milliseconds since the Unix epoch for a `Date` (matches C# DateTimeOffset.ToUnixTimeMilliseconds).
private func unixMillis(_ date: Date) -> Int64 {
    Int64((date.timeIntervalSince1970 * 1000).rounded())
}

private func encodeSpaceBreadcrumbWire(_ b: SpaceBreadcrumb) -> Data {
    // STANDARD base64; empty signature → "" (Data().base64EncodedString() already yields "").
    let sig = b.signature.base64EncodedString()
    let json = "{\"content_hash\":\(jsonEscaped(b.contentHash)),"
        + "\"geo_hash\":\(jsonEscaped(b.geoHash)),"
        + "\"anchor_uhid\":\(jsonEscaped(b.anchorUhid)),"
        + "\"created_at_ms\":\(unixMillis(b.createdAtUtc)),"
        + "\"ttl_hours\":\(b.ttlHours),"
        + "\"type\":\(Int(b.type.rawValue)),"
        + "\"signature\":\(jsonEscaped(sig))}"
    return Data(json.utf8)
}

private func parseSpaceBreadcrumbWire(_ data: Data) -> SpaceBreadcrumb? {
    guard let w = try? JSONDecoder().decode(SpaceBreadcrumbWire.self, from: data) else { return nil }
    let signature = Data(base64Encoded: w.signature) ?? Data()
    return SpaceBreadcrumb(
        contentHash: w.content_hash,
        geoHash: w.geo_hash,
        anchorUhid: w.anchor_uhid,
        createdAtUtc: Date(timeIntervalSince1970: Double(w.created_at_ms) / 1000.0),
        ttlHours: w.ttl_hours,
        type: BreadcrumbType(rawValue: UInt8(truncatingIfNeeded: w.type)) ?? .notice,
        signature: signature
    )
}

/// Test-only shim exposing the real ``SpaceBreadcrumb`` wire encoder (the encoder stays
/// `private`) so byte-identity vectors in `fixtures/space/vectors.json` can be verified.
internal func _spaceBreadcrumbWireBytesForTests(_ b: SpaceBreadcrumb) -> Data {
    encodeSpaceBreadcrumbWire(b)
}
